File: apps/synapse-web/src/domain/entities/widget.ts
export type ChartType =
	| "kpi"
	| "line"
	| "bar"
	| "pie"
	| "donut"
	| "area"
	| "stacked-bar"
	| "grouped-bar"
	| "scatter"
	| "bubble"
	| "heatmap"
	| "gauge"
	| "progress"
	| "table"
	| "sparkline"
	| "funnel"
	| "radar"
	| "treemap"
	| "waterfall"
	| "bullet";

export type WidgetCategory =
	| "finance"
	| "operations"
	| "research"
	| "manufacturing"
	| "quality"
	| "sales"
	| "supply-chain"
	| "hr";

export interface KPIData {
	value: number | string;
	change?: number;
	changeLabel?: string;
	secondaryValue?: string;
}

export interface TimeSeriesDataPoint {
	date: string;
	value: number;
	series?: string;
}

export interface CategoryDataPoint {
	label: string;
	value: number;
	color?: string;
	percentage?: number;
}

export interface TableData {
	columns: Array<{ key: string; label: string; type?: "text" | "number" | "currency" | "date" | "status" }>;
	rows: Array<Record<string, unknown>>;
}

export interface GaugeData {
	value: number;
	min: number;
	max: number;
	target?: number;
	thresholds?: Array<{ value: number; color: string; label: string }>;
}

export interface FunnelData {
	stages: Array<{ label: string; value: number; color?: string }>;
}

export interface HeatmapData {
	xLabels: string[];
	yLabels: string[];
	values: number[][];
}

export interface RadarData {
	axes: Array<{ label: string; value: number; maxValue?: number }>;
	series?: Array<{ name: string; values: number[] }>;
}

export interface TreemapData {
	items: Array<{
		label: string;
		value: number;
		color?: string;
		children?: Array<{ label: string; value: number; color?: string }>;
	}>;
}

export interface WaterfallData {
	items: Array<{
		label: string;
		value: number;
		type: "positive" | "negative" | "total";
	}>;
}

export interface ScatterData {
	points: Array<{ x: number; y: number; label?: string; series?: string; size?: number }>;
}

export interface WidgetDataMap {
	kpi: KPIData;
	line: TimeSeriesDataPoint[];
	bar: CategoryDataPoint[];
	pie: CategoryDataPoint[];
	donut: CategoryDataPoint[];
	area: TimeSeriesDataPoint[];
	"stacked-bar": Array<{ label: string; series: string; value: number }>;
	"grouped-bar": Array<{ label: string; series: string; value: number }>;
	scatter: ScatterData;
	bubble: ScatterData;
	heatmap: HeatmapData;
	gauge: GaugeData;
	progress: { value: number; max: number; label?: string };
	table: TableData;
	sparkline: number[];
	funnel: FunnelData;
	radar: RadarData;
	treemap: TreemapData;
	waterfall: WaterfallData;
	bullet: { value: number; target: number; ranges: Array<{ min: number; max: number; color: string }> };
}

File: apps/synapse-api/synapse_api/presentation/schemas/widget.py
from pydantic import BaseModel, Field
from typing import Any, Literal
from enum import Enum


class ChartType(str, Enum):
    KPI = "kpi"
    LINE = "line"
    BAR = "bar"
    PIE = "pie"
    DONUT = "donut"
    AREA = "area"
    STACKED_BAR = "stacked-bar"
    GROUPED_BAR = "grouped-bar"
    SCATTER = "scatter"
    BUBBLE = "bubble"
    HEATMAP = "heatmap"
    GAUGE = "gauge"
    PROGRESS = "progress"
    TABLE = "table"
    SPARKLINE = "sparkline"
    FUNNEL = "funnel"
    RADAR = "radar"
    TREEMAP = "treemap"
    WATERFALL = "waterfall"
    BULLET = "bullet"


class WidgetCategory(str, Enum):
    FINANCE = "finance"
    OPERATIONS = "operations"
    RESEARCH = "research"
    MANUFACTURING = "manufacturing"
    QUALITY = "quality"
    SALES = "sales"
    SUPPLY_CHAIN = "supply-chain"
    HR = "hr"


class WidgetConfig(BaseModel):
    id: str
    type: ChartType
    title: str
    category: WidgetCategory
    dataKey: str
    params: dict[str, Any] = Field(default_factory=dict)
    gridPosition: dict[str, int] = Field(
        default_factory=lambda: {"x": 0, "y": 0, "w": 6, "h": 4, "minW": 4, "minH": 3}
    )


class WidgetDataRequest(BaseModel):
    widgets: list[WidgetConfig]


class KPIData(BaseModel):
    value: int | float | str
    change: float | None = None
    changeLabel: str | None = None
    secondaryValue: str | None = None


class TimeSeriesDataPoint(BaseModel):
    date: str
    value: float
    series: str | None = None


class CategoryDataPoint(BaseModel):
    label: str
    value: float
    color: str | None = None
    percentage: float | None = None


class TableColumn(BaseModel):
    key: str
    label: str
    type: Literal["text", "number", "currency", "date", "status"] = "text"


class TableData(BaseModel):
    columns: list[TableColumn]
    rows: list[dict[str, Any]]


class GaugeData(BaseModel):
    value: float
    min: float
    max: float
    target: float | None = None
    thresholds: list[dict[str, Any]] | None = None


class FunnelStage(BaseModel):
    label: str
    value: float
    color: str | None = None


class FunnelData(BaseModel):
    stages: list[FunnelStage]


class HeatmapData(BaseModel):
    xLabels: list[str]
    yLabels: list[str]
    values: list[list[float]]


class RadarAxis(BaseModel):
    label: str
    value: float
    maxValue: float | None = None


class RadarData(BaseModel):
    axes: list[RadarAxis]
    series: list[dict[str, Any]] | None = None


class TreemapItem(BaseModel):
    label: str
    value: float
    color: str | None = None
    children: list[dict[str, Any]] | None = None


class TreemapData(BaseModel):
    items: list[TreemapItem]


class WaterfallItem(BaseModel):
    label: str
    value: float
    type: Literal["positive", "negative", "total"]


class WaterfallData(BaseModel):
    items: list[WaterfallItem]


class ScatterPoint(BaseModel):
    x: float
    y: float
    label: str | None = None
    series: str | None = None
    size: float | None = None


class ScatterData(BaseModel):
    points: list[ScatterPoint]


class BulletData(BaseModel):
    value: float
    target: float
    ranges: list[dict[str, Any]]


class ProgressData(BaseModel):
    value: float
    max: float
    label: str | None = None


class WidgetDataResponse(BaseModel):
    data: dict[str, Any]


class DashboardDataRequest(BaseModel):
    widgets: list[WidgetConfig]


class DashboardDataResponse(BaseModel):
    widgets: dict[str, Any]
